src/main/java/io/aiven/kafka/auth/AivenAclAuthorizerV2.java

```diff
@@ -205,13 +210,170 @@ public final List<AuthorizationResult> authorize(final AuthorizableRequestContex
+            new EnumMap<PatternType, Set<String>>(PatternType.class) {
+                {
+                    put(PatternType.LITERAL, new HashSet<>());
+                    put(PatternType.PREFIXED, new HashSet<>());
+                }
+            };
+        final EnumMap<PatternType, Set<String>> allowPatterns =
+            new EnumMap<PatternType, Set<String>>(PatternType.class) {
+                {
+                    put(PatternType.LITERAL, new HashSet<>());
+                    put(PatternType.PREFIXED, new HashSet<>());
+                }
+            };
+
+        final String hostAddr = requestContext.clientAddress().getHostAddress();
+
+        for (final AivenAcl acl : this.cacheReference.get().getDenyAclEntries()) {
+
+            if (!acl.hostMatch(hostAddr)) {
+                continue;
+            }
+
+            if (!acl.matchResourceType(resourceType)) {
+                continue;
+            }
+
+            if (!acl.matchPrincipal(principalType, principalName)) {
+                continue;
+            }
+
+            if (!acl.matchOperation(op)) {
+                continue;
+            }
+
+            for (final AclBinding binding : AclAivenToNativeConverter.convert(acl)) {
+
+                if (binding.pattern().patternType() == PatternType.LITERAL) {
+                    // If wildcard deny exists, return deny directly
+                    if (binding.pattern().name().equals(ResourcePattern.WILDCARD_RESOURCE)) {
+                        return AuthorizationResult.DENIED;
+                    }
+                    denyPatterns.get(PatternType.LITERAL).add(binding.pattern().name());
+                } else if (binding.pattern().patternType() == PatternType.PREFIXED) {
+                    denyPatterns.get(PatternType.PREFIXED).add(binding.pattern().name());
+                }
+            }
+        }
+
+        for (final AivenAcl acl : this.cacheReference.get().getAllowAclEntries()) {
+
+            if (!acl.hostMatch(hostAddr)) {
+                continue;
+            }
+
+            if (!acl.matchResourceType(resourceType)) {
+                continue;
+            }
+
+            if (!acl.matchPrincipal(principalType, principalName)) {
+                continue;
+            }
+
+            if (!acl.matchOperation(op)) {
+                continue;
+            }
+
+            for (final AclBinding binding : AclAivenToNativeConverter.convert(acl)) {
+
+                if (binding.pattern().patternType() == PatternType.LITERAL) {
+                    if (binding.pattern().name().equals(ResourcePattern.WILDCARD_RESOURCE)) {
+                        return AuthorizationResult.ALLOWED;
+                    }
+                    allowPatterns.get(PatternType.LITERAL).add(binding.pattern().name());
+                } else if (binding.pattern().patternType() == PatternType.PREFIXED) {
+                    allowPatterns.get(PatternType.PREFIXED).add(binding.pattern().name());
+                }
```

Comment:
What happens on fallthrough here, maybe raise and log an error as it would be unexpected?

Reply:
Those types are statically defined, and `AclAivenToNativeConverter.convert` should not return anything else. I think raising an error here would not be desired and safe. However if somehow the code is later extended (and Kafka adds support for something else), then adding an error log makes sense? Sounds good?

Reply:
I can live with the error logging, my concern is that if on fallthrough access would be granted when it should not.

Reply:
Added logging for both allow and deny loops, but only returning denied from deny loop. If something skipped from allow loop, then allowing by other rules is fine in my logic.